internal/safefile/safe_object_directory.go

```diff
@@ -0,0 +1,99 @@
+package safefile
+
+import (
+	"path/filepath"
+	"syscall"
+	"unicode/utf16"
+	"unsafe"
+)
+
+const globalNTPath = "\\Global??"
+const bufferSize = 1024
+
+func ntWidePath(path string) ([]uint16, error) {
+	path = filepath.Clean(path)
+	fspath := filepath.FromSlash(path)
+	path16 := utf16.Encode(([]rune)(fspath))
+	if len(path16) > 32767 {
+		return nil, syscall.ENAMETOOLONG
+	}
+	return path16, nil
+}
+
+func EnumerateNTGlobalObjectDirectory() ([]string, error) {
+	var (
+		handle uintptr
+		oa     objectAttributes
+	)
+
+	globalNTPath16, err := ntWidePath(globalNTPath)
+	if err != nil {
+		return nil, err
+	}
+
+	upathBuffer := localAlloc(0, int(unsafe.Sizeof(unicodeString{}))+len(globalNTPath16)*2)
+	upath := (*unicodeString)(unsafe.Pointer(upathBuffer))
+	upath.Length = uint16(len(globalNTPath16) * 2)
+	upath.MaximumLength = upath.Length
+	upath.Buffer = upathBuffer + unsafe.Sizeof(*upath)
+	copy((*[32768]uint16)(unsafe.Pointer(upath.Buffer))[:], globalNTPath16)
+	defer localFree(upathBuffer)
+
+	oa.Length = unsafe.Sizeof(oa)
+	oa.ObjectName = upathBuffer
+
+	status := ntOpenDirectoryObject(
+		&handle,
+		FILE_LIST_DIRECTORY,
+		&oa,
+	)
+
+	if !NTSuccess(status) {
+		return nil, rtlNtStatusToDosError(status)
+	}
+
+	defer syscall.Close(syscall.Handle(handle))
+
+	var (
+		context      uint32
+		returnLength uint32
+		buffer       [bufferSize]byte
+		result       []string
+	)
+	for {
+		status = ntQueryDirectoryObject(
+			handle,
+			&buffer[0],
+			bufferSize,
+			false,
+			false,
+			&context,
+			&returnLength,
+		)
+
+		if !NTSuccess(status) && status != _STATUS_MORE_ENTRIES {
+			if status == _STATUS_NO_MORE_ENTRIES || status == _ERROR_NO_MORE_ITEMS {
+				break
+			}
+			return nil, rtlNtStatusToDosError(status)
+		}
+		dirInfo := (*objectDirectoryInformation)(unsafe.Pointer(&buffer[0]))
+		index := 0
+		for {
+			index++
```

Comment:
Feels weird to have the index start at 0 if we just increment it before the initial value ever gets used. Why not start at 1 and increment at the end of this nested for.